### PR TITLE
fix: bound DSA prefill score memory

### PR DESCRIPTION
**Pull request by CUHKSZzxy**
Supersedes #4892 by carrying its sparse FlashMLA refactor together with the DSA memory and short-prefill optimizations.

## Summary

- Separate dense MLA and sparse DSA FlashMLA execution, with implementation selection owned by the CUDA attention builder.
- Bound DeepGEMM DSA prefill scores by chunking query rows under a configurable 512 MiB budget, reusing the flattened indexer K cache, and reserving the same workspace before KV-cache sizing, addressing the same `[M, N]` FP32 logits OOM as [vLLM #36178](https://github.com/vllm-project/vllm/pull/36178).
- When the full prefill context fits within DSA top-k, skip index scoring and sparse mapping and use equivalent dense attention, including FA3 when available. Indexer K-cache writes are preserved, while MTP still computes the indices required by later draft iterations.
- Default DSA MLA KV cache to BF16; an explicit FP8 cache policy still selects the packed FP8 MLA layout.

## Validation

- Pre-commit checks passed.
- Focused attention, DSA indexer, MTP, configuration, and cache-engine tests passed (`60 passed`).
- GLM-5.2 TP8 SWE serving completed `80/80` requests with FP8 KV cache, MTP5, prefix caching, and symmetric-memory all-reduce: `914.67 tok/s` overall output and `1425.27 tok/s` steady completion throughput.
- The SWE prompts use 16K+ contexts, so this run validates the existing long-context sparse path and absence of a serving regression; it does not measure the short-prefill dense-routing speedup.

## Assistance

Assisted with Codex + GPT-5.6-Sol xHigh, reviewed manually

### CODE REVIEW

File: lmdeploy/pytorch/backends/cuda/attention/sparse_mla.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        if attn_metadata.max_kv_seqlen <= self.mla_index_topk:
+            # Top-k contains every valid key, so dense attention is equivalent
+            # and avoids sparse index mapping and kernel overhead.
+            return super()._forward_prefill(query,
+                                            k_cache,
+                                            v_cache,
+                                            attn_metadata,
+                                            nsa_indices=None,
+                                            k_scales_zeros=k_scales_zeros,
+                                            v_scales_zeros=v_scales_zeros)
```
[the remark this answers is not in the document]
Comment (CUHKSZzxy) [reply lmdeploy/pytorch/backends/cuda/attention/sparse_mla.py:239]: Fixed in bd068e63. When BF16 MLA flattening returns an empty V cache, the dense Triton prefill fallback now uses a zero-copy view of the leading c_kv payload in flatten_k, restoring the required 512-wide value input. Added regression coverage.

File: lmdeploy/pytorch/backends/cuda/nsa.py [hunk 2]
```diff
@@ -292,84 +323,134 @@ def get_step_metadata(self, attn_metadata) -> NSAIndexMeta:
...
+            self, q: Tensor, q_s: Tensor, indexer_k_cache: Tensor,
+            meta: NSAIndexMeta,
+            score_meta: _DeepGemmContiguousScoreMeta) -> Tensor:
+        """Bound prefill score memory by chunking only the query rows."""
+        # Keep KV contiguous for DeepGEMM's TMA descriptors; slicing only Q
+        # avoids the alignment failures caused by per-request KV views.
+        flat_k, flat_k_s = self._flatten_prefill_k(
+            indexer_k_cache, q.size(-1), meta)
+        max_rows = _get_max_score_rows(score_meta.max_kv_seqlen,
+                                       self.max_logits_bytes)
```
[the remark this answers is not in the document]
Comment (CUHKSZzxy) [reply lmdeploy/pytorch/backends/cuda/nsa.py:443]: No additional change is needed here. Supported DSA top-k widths (512 and 2048) dispatch to sparse_index_topk, and TileLang is a CUDA runtime dependency. bitonic_topk remains only a compatibility fallback, so the bounded production path does not allocate these full-width temporaries.

File: lmdeploy/pytorch/backends/cuda/nsa.py [hunk 3]
```diff
@@ -292,84 +323,134 @@ def get_step_metadata(self, attn_metadata) -> NSAIndexMeta:
...
+            del scores, selected
+        return out
+
     def _score_and_select(self, q: Tensor, q_s: Tensor,
                           indexer_k_cache: Tensor, meta: NSAIndexMeta) -> Tensor:
-        scores = self._compute_scores(q, q_s, indexer_k_cache, meta)
+        score_meta = meta.score_meta
+        if isinstance(score_meta, _DeepGemmContiguousScoreMeta):
+            return self._score_and_select_prefill(
+                q, q_s, indexer_k_cache, meta, score_meta)
```
[the remark this answers is not in the document]
Comment (CUHKSZzxy) [reply lmdeploy/pytorch/backends/cuda/nsa.py:471]: Fixed in bd068e63. Before the Triton fallback allocates its full [rows, max_kv_seqlen] FP32 score tensor, it now checks the exact allocation size and raises an explicit DeepGEMM-required error when it exceeds the configured limit. Added regression coverage.

**Review by grimoire — APPROVED** [review]
LGTM
